src/sources/kafka.rs

```diff
@@ -218,6 +233,99 @@ async fn kafka_source(
     Ok(())
 }
 
+#[derive(Debug)]
+struct CommitterEntry {
+    done: bool,
+    topic: String,
+    partition: i32,
+    offset: i64,
+}
+
+impl<'a> From<BorrowedMessage<'a>> for CommitterEntry {
+    fn from(msg: BorrowedMessage<'a>) -> Self {
+        Self {
+            done: false,
+            topic: msg.topic().into(),
+            partition: msg.partition(),
+            offset: msg.offset(),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct CommitterStore {
+    next: u64,
+    map: HashMap<u64, CommitterEntry>,
+}
+
+impl CommitterStore {
+    fn extract_topic_partition_list(&mut self) -> Option<TopicPartitionList> {
+        // Collect all the message offsets that can be stored by this one being finalied.
+        let mut next = self.next;
+        let mut result = None;
+        while let Entry::Occupied(entry) = self.map.entry(next) {
+            if !entry.get().done {
+                break;
+            }
+            let entry = entry.remove();
+            result
+                .get_or_insert_with(TopicPartitionList::new)
+                .add_partition(&entry.topic, entry.partition)
+                .set_offset(Offset::from_raw(entry.offset + 1)) // FIXME why does the +1 make it work?
```

Comment:
When I just use `Offset::from_raw(entry.offset)`, this fails with an "invalid parameter" error. Adding the `+ 1` makes it work, and produces the same high water mark as the non-acknowledgement code path. I have no idea why.